### PR TITLE
Clean up and reorganize annotation scanning code

### PR DESCRIPTION
**Pull request by coderbot16**
This should make the logic a bit more clear and a bit less fragile.

### CODE REVIEW

**Comment by TheGlitch76** [comment]
Requesting a review from @qouteall

File: patchwork-fml/src/main/java/net/minecraftforge/fml/ModList.java [hunk 1]
```diff
@@ -53,73 +53,69 @@ public boolean isLoaded(String modId) {
...
-
-		if (source == null) {
-			return false;
+	public List<ModFileScanData> getAllScanData() {
+		if (allScanDataCache == null) {
+			allScanDataCache = FabricLoader.getInstance().getAllMods()
+					.stream()
+					.map(modContainer -> modContainer.getMetadata().getId())
+					.map(modid -> getModFileById(modid).getFile().getScanResult())
+					.distinct()
```
Comment (TheGlitch76) [inline patchwork-fml/src/main/java/net/minecraftforge/fml/ModList.java:115]: ModFileScanData doesn't implement hashcode or equals so this does nothing.

File: patchwork-fml/src/main/java/net/minecraftforge/forgespi/language/ModFileScanData.java [hunk 2]
```diff
@@ -66,24 +66,14 @@ private ModFileScanData() {
...
 		Path annotationJsonPath = modContainer.getPath(annotationJsonLocation);
 
 		try {
-			InputStream outputStream = Files.newInputStream(annotationJsonPath);
-			Gson gson = new Gson();
-			this.annotationStorage = gson.fromJson(
-					new InputStreamReader(outputStream), AnnotationStorage.class
-			);
-			annotationData = this.annotationStorage.entries.stream()
+			InputStreamReader reader = new InputStreamReader(Files.newInputStream(annotationJsonPath));
```
Comment (TheGlitch76) [inline patchwork-fml/src/main/java/net/minecraftforge/forgespi/language/ModFileScanData.java:72]: is this inputstream properly closed?

File: patchwork-fml/src/main/java/net/minecraftforge/fml/ModList.java (hunk 1 above)
Comment (coderbot16) [reply patchwork-fml/src/main/java/net/minecraftforge/fml/ModList.java:115]: In this case there is only one instance of ModFileScanData for a given mod file (mod files can be shared by multiple mod containers), therefore comparison by object identity alone (`==`) is sufficient. I'll add a comment for this.

File: patchwork-fml/src/main/java/net/minecraftforge/fml/ModList.java [hunk 3]
```diff
@@ -108,12 +109,17 @@ private ModFileInfo createModFileInfo(ModContainer modContainer) {
 
 	public List<ModFileScanData> getAllScanData() {
 		if (allScanDataCache == null) {
-			allScanDataCache = FabricLoader.getInstance().getAllMods()
+			// Even though ModFIleScanData lacks an implementation of Object#equals, the default implementation tests
```
Comment (TheGlitch76) [inline patchwork-fml/src/main/java/net/minecraftforge/fml/ModList.java:112]:
```suggestion
			// Even though ModFileScanData lacks an implementation of Object#equals, the default implementation tests
```

**Comment by TheGlitch76** [comment]
oops

**Comment by qouteall** [comment]
ok

**Pull request merged**